protocol/buff/buff.go

```diff
@@ -0,0 +1,288 @@
+// for the previous message.
+func (b *Buff) BeginMessage(mType uint8) {
+	if b.mPos > 0 {
+		panic("cannot begin message: the previous message is not finished")
+	}
+
+	b.mPos = 1 + len(b.payload)
+	b.payload = append(b.payload, mType, 0, 0, 0, 0)
+}
+
+// EndMessage sets the `message_length` allocated by BeginMessage.
+// EndMessage panics if BeginMessage was not called first
+// or if BeginBytes was not followed by EndBytes.
+func (b *Buff) EndMessage() {
+	// todo check unfinished bytes
+	if b.mPos <= 0 {
+		panic("cannot end message: no current message")
+	}
+
+	if len(b.bPos) > 0 {
+		panic("cannot end message: bytes in progress")
+	}
+
+	length := uint32(len(b.payload) - b.mPos)
+	binary.BigEndian.PutUint32(b.payload[b.mPos:b.mPos+4], length)
+	b.mPos = 0
+}
+
+// Unwrap returns a pointer to the buffers *[]byte.
+func (b *Buff) Unwrap() *[]byte {
+	if b.mPos > 0 {
+		panic("cannot unwrap: the previous message is not finished")
+	}
+
+	return &b.payload
+}
+
+// Next returns true if the buffer is not fully read.
+func (b *Buff) Next() bool {
+	b.Finish()
+
+	if b.mPos < len(b.payload) {
+		if len(b.payload) < b.mPos+5 {
+			panic("buffer overread")
+		}
+
+		pos := 1 + int(binary.BigEndian.Uint32(b.payload[b.mPos+1:b.mPos+5]))
+
+		if len(b.payload) < b.mPos+pos {
+			panic("buffer overread")
+		}
+
+		b.Msg = b.payload[b.mPos+5 : b.mPos+pos]
+		b.msgHdr.Cap = b.msgHdr.Len
+		b.MsgType = b.payload[b.mPos]
+		b.mPos += pos
+
+		return true
+	}
+
+	b.msgHdr.Cap = 0
+	b.msgHdr.Len = 0
+	b.MsgType = 0
+	return false
+}
+
+// Finish asserts that the message has been fully read.
+// It panics if it has not.
+func (b *Buff) Finish() {
+	if len(b.Msg) > 0 {
+		panic(fmt.Sprintf(
+			"cannot finish: unread data in buffer (message type: 0x%x)",
+			b.MsgType,
+		))
+	}
+}
+
+// Len returns the number of bytes remaining to be read.
+func (b *Buff) Len() int {
+	return len(b.Msg)
+}
+
+// AssertAllocated panics if there aren't n bytes in the buffer.
+func (b *Buff) AssertAllocated(n int) {
+	if len(b.Msg) < n {
+		panic("buffer overread")
+	}
+}
+
+// Discard skips the next n bytes.
+func (b *Buff) Discard(n int) {
+	b.AssertAllocated(n)
+	b.Msg = b.Msg[n:]
+}
+
+// PopUint8 returns the next byte and advances the buffer.
+func (b *Buff) PopUint8() uint8 {
+	val := b.Msg[0]
+	b.Msg = b.Msg[1:]
+	return val
+}
+
+// PopUint16 reads a uint16 and advances the buffer.
+func (b *Buff) PopUint16() uint16 {
+	val := binary.BigEndian.Uint16(b.Msg)
+	b.Msg = b.Msg[2:]
+	return val
+}
+
+// PopUint32 reads a uint32 and advances the buffer.
+func (b *Buff) PopUint32() uint32 {
+	val := b.PeekUint32()
+	b.Msg = b.Msg[4:]
+	return val
+}
+
+// PeekUint32 reads a uint32 but does not advance the buffer.
+func (b *Buff) PeekUint32() uint32 {
+	return binary.BigEndian.Uint32(b.Msg)
+}
+
+// PopUint64 reads a uint64 and advances the buffer.
+func (b *Buff) PopUint64() uint64 {
+	val := binary.BigEndian.Uint64(b.Msg[:8])
```

Comment:
Will it complain if `b.Msg` has less than 8 bytes in it?

Reply:
yes it will panic

Reply:
Good.